### PR TITLE
feat: terragrunt support

### PR DESCRIPTION
**Pull request by bgdanix**
#5283  Changes:

Allows for checking of Terragrunt config files in order to update the Terraform module version used inside them

## Context:

Terragrunt uses the same logic as for Terraform modules, with the same `source = "<git-url>"` format, so we can also have the capability of Terragrunt updates. Example of a Terragrunt .hcl config file:

```hcl
terraform {
  source = "github.com/githubuser/repo//terraform/modules/mymodule?ref=v1.0.5"
}
```
## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added unit tests, or
- [x] Unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — CHANGES_REQUESTED** [review]
I'd like to see a test with a "real"/full terragrunt file used as a fixture

**Comment by bgdanix** [comment]
Added some tests for this scenario. I would appreciate id you could help me to make the pipelines pass, it's failing due to global coverage. Thank you

**Comment by rarkins** [comment]
@bgdanix if you run `yarn jest manager/terraform` locally then open `coverage/lib/manager/terraform/index.html` then you can see this:

image

You don't need to worry about `index.ts` (it's loaded/covered by other tests) however the missing coverage in the other files seems caused by changes in this PR.

**Comment by bgdanix** [comment]
> @bgdanix if you run `yarn jest manager/terraform` locally then open `coverage/lib/manager/terraform/index.html` then you can see this:

Sorry but I can't figure out what to do, especially since I haven't touched some of those files... which to me does not make sense to have such low coverage considering nothing changed.
I did run renovate against a real test repo and the functionality implemented works as expected.

**Comment by rarkins** [comment]
PRs cannot be merged if they reduce coverage. Even if it's not code you directly modified then it means something about your new code or tests has changed the execution path of existing tests and things aren't being tested properly as a result.

File: lib/manager/terraform/extract.spec.ts [hunk 1]
```diff
@@ -16,8 +21,8 @@ describe('lib/manager/terraform/extract', () => {
     it('extracts', () => {
       const res = extractPackageFile(tf1);
       expect(res).toMatchSnapshot();
-      expect(res.deps).toHaveLength(44);
-      expect(res.deps.filter((dep) => dep.skipReason)).toHaveLength(8);
+      expect(res.deps).toHaveLength(39);
```
Comment (rarkins) [inline lib/manager/terraform/extract.spec.ts:24]: Why is this changed? Adding functionality shouldn't normally decrease the number of dependencies detected

File: lib/manager/terraform/__snapshots__/extract.spec.ts.snap [hunk 2]
```diff
@@ -280,70 +280,13 @@ Object {
       "lookupName": "ssh://git@bitbucket.com/hashicorp/example",
     },
     Object {
-      "currentValue": ">= 2.7.0",
-      "datasource": "terraform-provider",
-      "depName": "aws",
-      "depNameShort": "aws",
-      "depType": "terraform",
+      "skipReason": "no-source",
```
Comment (rarkins) [inline lib/manager/terraform/__snapshots__/extract.spec.ts.snap:283]: Why would these examples no longer work?

**Review by rarkins — CHANGES_REQUESTED** [review]
The decrease in coverage appears related to a breaking of tests too

File: lib/manager/terraform/extract.spec.ts (hunk 1 above)
Comment (bgdanix) [reply lib/manager/terraform/extract.spec.ts]: Indeed. After some more debugging I realized using the same terraform manager to cover terragrunt is not going to work, so I'm working on setting up a new manager.

File: lib/manager/terraform/util.ts [hunk 3]
```diff
@@ -55,6 +55,9 @@ export function getTerraformDependencyType(
     case 'module': {
       return TerraformDependencyTypes.module;
     }
+    case 'terraform': {
+      return TerraformDependencyTypes.module;
+    }
```
Comment (rarkins) [inline lib/manager/terraform/util.ts:60]: These three lines cause the existing extract test to fail. It seems to be matching some of the existing extracted dependencies.

**Comment by rarkins** [comment]
Here's the status:

1. I've reverted the test case edits and snapshot changes for existing tests. These should stay as they are, unless it can be shown that they were previously *wrong*

2. I've identified which new line is causing the existing cases to fail. @bgdanix please take a look at that and work out if it's intentional or not. It seems to decrease the number of detected dependencies which I assume is not intentional

File: lib/manager/terraform/util.ts (hunk 3 above)
Comment (bgdanix) [reply lib/manager/terraform/util.ts]: No longer applies. Have created a new manager

**Comment by bgdanix** [comment]
In the end I had to create a new manager called `terragrunt` as it has quite some overlapping with Terragrunt manager and it would have required a lot of rework. As you see now all pipelines have passed and I've also ran it on a close to reality example as it can be [seen here](https://github.com/bgdanix/renovate-tst/pull/4)

**Review by viceice — COMMENTED** [review]
Can we share some common code with Terraform manager instead of duplicate it?

**Comment by bgdanix** [comment]
> 
> Can we share some common code with Terraform manager instead of duplicate it?
Actually there's nothing to share with Terraform manager, all the functions are different and have a slightly different logic

**Comment by bgdanix** [comment]
@viceice , @rarkins are you ok with this? Can it be approved?

File: lib/manager/terragrunt/readme.md [hunk 4]
```diff
@@ -0,0 +1,9 @@
+Currently Terragrunt support is limited to Terraform registry sources and github sources that include semver refs, e.g. like `github.com/hashicorp/example?ref=v1.0.0`.
```
Comment (rarkins) [inline lib/manager/terragrunt/readme.md:1]: I'm interested to know how many other types there are which we might want to support in future?

File: lib/manager/terragrunt/index.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import * as hashicorpVersioning from '../../versioning/hashicorp';
+
+export { extractPackageFile } from './extract';
+
+export const defaultConfig = {
+  commitMessageTopic:
+    'Terragrunt {{managerData.terragruntDependencyType}} {{depNameShort}}',
```
Comment (rarkins) [inline lib/manager/terragrunt/index.ts:7]: Won't this result in the commit messages and PR titles being a bit awkward? e.g. "Update Terragrunt gitTags foo/bar". Would "Terragrunt module" or "Terragrunt dependency" work better?

Comment (bgdanix) [reply lib/manager/terragrunt/index.ts:7]: Agree, updated to reflect that. See this PR: https://github.com/bgdanix/renovate-tst/pull/7

File: lib/manager/terragrunt/readme.md (hunk 4 above)
Comment (bgdanix) [reply lib/manager/terragrunt/readme.md]: I've updated documentation to include info about custom config for using ref which are not in line with semver.

The text you mentioned is basically the same as for Terraform, but just FYI all the available source types supported by both of them can be [found here](https://www.terraform.io/docs/modules/sources.html)

File: lib/manager/terragrunt/readme.md [hunk 6]
```diff
@@ -0,0 +1,17 @@
+Currently by default, Terragrunt support is limited to Terraform registry sources and github sources that include semver refs, e.g. like `github.com/hashicorp/example?ref=v1.0.0`.
+
+It is however possible to also use custom [versioning config](../../../docs/usage/configuration-options.md#versioning) in order to support non-semver references. For example, if your want to ref a tag like `module-v1.2.5`, a block like the following would solve this issue:
```
Comment (HonkingGoose) [inline lib/manager/terragrunt/readme.md:3]:
```suggestion
It is however possible to also use custom [versioning config](../../../docs/usage/configuration-options.md#versioning) in order to support non-semver references.
For example, if your want to ref a tag like `module-v1.2.5`, a block like the following would solve this issue:
```

This suggestion makes it so we conform to the one sentence per line style.

File: lib/manager/terragrunt/readme.md [hunk 7]
```diff
@@ -0,0 +1,18 @@
+Currently by default, Terragrunt support is limited to Terraform registry sources and github sources that include semver refs, e.g. like `github.com/hashicorp/example?ref=v1.0.0`.
```
Comment (HonkingGoose) [inline lib/manager/terragrunt/readme.md:1]:
```suggestion
Currently by default, Terragrunt support is limited to Terraform registry sources and GitHub sources that include semver refs, e.g. like `github.com/hashicorp/example?ref=v1.0.0`.
```

GitHub is a noun and should be capitalized.

File: lib/manager/terragrunt/readme.md [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+It is however possible to also use custom [versioning config](../../../docs/usage/configuration-options.md#versioning) in order to support non-semver references.
+For example, if your want to ref a tag like `module-v1.2.5`, a block like the following would solve this issue:
```
Comment (HonkingGoose) [inline lib/manager/terragrunt/readme.md:4]:
```suggestion
For example, if you want to reference a tag like `module-v1.2.5`, a block like this would work:
```

Rewrite sentence for clarity.

File: lib/manager/terragrunt/readme.md [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+It is however possible to also use custom [versioning config](../../../docs/usage/configuration-options.md#versioning) in order to support non-semver references.
+For example, if you want to reference a tag like `module-v1.2.5`, a block like this would work:
+
+```json
+"terraform": {
+	"versioning": "regex:^((?<compatibility>.*)-v|v*)(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)$"
+}
+```
+
+Fixed Terragrunt versions like the following will receive a PR whenever there is a newer version available:
```
Comment (HonkingGoose) [inline lib/manager/terragrunt/readme.md:12]: It's not clear to me what this "Fixed Terragrunt versions" is talking about.
The word fixed brings to mind things like: "Oh it was broken, but now it works." or "I'm hard-coding a version here".
I don't know how to improve this sentence.

Maybe somebody else has a good suggestion for this sentence?

File: lib/manager/terragrunt/readme.md [hunk 10]
```diff
@@ -0,0 +1,18 @@
+Currently by default, Terragrunt support is limited to Terraform registry sources and GitHub sources that include semver refs, e.g. like `github.com/hashicorp/example?ref=v1.0.0`.
+
+It is however possible to also use custom [versioning config](../../../docs/usage/configuration-options.md#versioning) in order to support non-semver references.
```
[the remark this answers is not in the document]
Comment (bgdanix) [reply lib/manager/terragrunt/readme.md:3]: Yep, shorter and concise

File: lib/manager/terragrunt/readme.md (hunk 9 above)
Comment (bgdanix) [reply lib/manager/terragrunt/readme.md:12]: @HonkingGoose the same is also in `terraform` manager. But anyway, to make it more clear I suggest this: `Pinned Terragrunt dependecies like the following will receive a PR whenever there is a newer version available:`

File: lib/manager/terragrunt/readme.md [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+"terraform": {
+	"versioning": "regex:^((?<compatibility>.*)-v|v*)(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)$"
+}
+```
+
+Fixed Terragrunt versions like the following will receive a PR whenever there is a newer version available:
+
+```hcl
+terraform {
+  source = "github.com/hashicorp/example?ref=v1.0.0"
```
Comment (HonkingGoose) [inline lib/manager/terragrunt/readme.md:16]: At line 1 in this document you say: "GitHub sources that include semver refs, e.g. like `github.com/hashicorp/example?ref=v1.0.0`."

But why do we use `source = "github.com/hashicorp/example?ref=v1.0.0"` in the "Fixed Terragrunt versions" example then? Are you sure that makes sense?

It's a bit confusing to me, to first say "Oh this works by default", and then later in a an code block say: "Oh wait you need to fix this by doing this thing".

File: lib/manager/terragrunt/readme.md (hunk 9 above)
Comment (HonkingGoose) [reply lib/manager/terragrunt/readme.md:12]: Hmm, I'm not sure if the word "pinned" is better. The word pinned is usually used to mean: "Use one this specific version", at least the way dependency management bots use the word like this.

File: lib/manager/terragrunt/readme.md (hunk 11 above)
Comment (bgdanix) [reply lib/manager/terragrunt/readme.md:16]: Sorry but I don't understand what the problem is. It's same GitHub source like in the example.

Comment (bgdanix) [reply lib/manager/terragrunt/readme.md:16]: Furthermore, people using terragrunt or terraform need to know how to use these tools before using renovate.

Comment (HonkingGoose) [reply lib/manager/terragrunt/readme.md:16]: Let's "resolve" this then, you're the maker of the code, so you know best if the docs are correct. :wink: 

Can you click the "resolve" button on this conversation? I don't have the rights to mark this as "resolved" myself. :smile:

File: lib/manager/terragrunt/readme.md (hunk 9 above)
Comment (bgdanix) [reply lib/manager/terragrunt/readme.md:12]: And that is the exact case here: you need to pin to a version (e.g use this specific version) in order for `renovate` to work and suggest new version available

File: lib/manager/terragrunt/readme.md (hunk 11 above)
Comment (bgdanix) [reply lib/manager/terragrunt/readme.md:16]: Sure thing :)

File: lib/manager/terragrunt/readme.md (hunk 9 above)
Comment (HonkingGoose) [reply lib/manager/terragrunt/readme.md:12]: > And that is the exact case here: you need to pin to a version (e.g use this specific version) in order for renovate to work and suggest new version available

---

```suggestion
Pinned Terragrunt dependencies like the the following will receive a PR whenever there is a newer version available:
```

Let's go with this then. :+1:
I fixed a small spelling error in your comment's suggestion with the word `dependencies`.

**Review by HonkingGoose — COMMENTED** [review]
The `readme.md` is good to go on my part. I'm not reviewing any other parts of this PR as I am not qualified to do so.

**Comment by bgdanix** [comment]
> 
> 
> The `readme.md` is good to go on my part. I'm not reviewing any other parts of this PR as I am not qualified to do so.

Thank you! @viceice , @rarkins good to go also from your side?

**Comment by HonkingGoose** [comment]
Dang it, I found "one more thing", I'm really sorry for this @bgdanix :disappointed: 

If you need to pin to `hashicorp/example/v1.0.0` do you then also need to bump to `hashicorp/example/v2.0.0` in the config file manually when you want to receive updates to a `v.2.0.0` version? The docs don't really mention this right now?

**Comment by bgdanix** [comment]
> 
> 
> Dang it, I found "one more thing", I'm really sorry for this @bgdanix 😞
> 
> If you need to pin to `hashicorp/example/v1.0.0` do you then also need to bump to `hashicorp/example/v2.0.0` in the config file manually when you want to receive updates to a `v2.0.0` version? The docs don't really mention this right now?

No, you don't. By default renovate will bump to the latest available version. So if right now you have v1.0.0 and the latest one available in the repo is v4.0.0 then it will raise a PR with this change. It's the user that has to decide if the PR should be merged on not. Later on, if a new version is available (let's say v6.0.0) then the process will be the same as the previous one.

**Comment by HonkingGoose** [comment]
> No, you don't.

:rofl: Whoops, this silly HonkingGoose will shut up now... :smile: :rofl: 
I've "resolved" my comment, feel free to proceed.

File: lib/manager/terragrunt/index.ts [hunk 12]
```diff
@@ -0,0 +1,10 @@
+import * as hashicorpVersioning from '../../versioning/hashicorp';
+
+export { extractPackageFile } from './extract';
+
+export const defaultConfig = {
+  commitMessageTopic:
+    'Terragrunt {{managerData.terragruntDependencyType}} dependecy {{depNameShort}}',
```
Comment (rarkins) [inline lib/manager/terragrunt/index.ts:7]:
```suggestion
    'Terragrunt dependency {{depNameShort}}',
```

Use `depType` if you must, but not `managerData`.

**Review by rarkins — CHANGES_REQUESTED** [review]
Do not extract `managerData`. Use `depType` if really necessary to use a category in the commit message / PR title

File: lib/manager/terragrunt/index.ts (hunk 12 above)
Comment (bgdanix) [reply lib/manager/terragrunt/index.ts:7]: Got it!

**Review by rarkins — DISMISSED** [review]
I think the code is too complex in parts which will make it challenging to maintain, however it's isolated within manager/terragrunt/* so less of a problem than if it were mixing with existing code. I'm OK to merge if @viceice is OK

File: lib/manager/terragrunt/index.ts [hunk 13]
```diff
@@ -0,0 +1,9 @@
+import * as hashicorpVersioning from '../../versioning/hashicorp';
+
+export { extractPackageFile } from './extract';
+
+export const defaultConfig = {
+  commitMessageTopic: 'Terragrunt dependency {{depNameShort}}',
+  fileMatch: ['\\.hcl$'],
```
Comment (viceice) [inline lib/manager/terragrunt/index.ts:7]:
```suggestion
  fileMatch: ['(^|/)terragrunt\\.hcl$'],
```

I think we should be more strict here as terragrunt only reads `terragrunt.hcl` files by default, see [docs](https://terragrunt.gruntwork.io/docs/getting-started/configuration/#terragrunt-configuration-file)

A user can always override this if required.

**Review by viceice — COMMENTED** [review]
A small change, otherwise LGTM

File: lib/manager/terragrunt/index.ts (hunk 13 above)
Comment (bgdanix) [reply lib/manager/terragrunt/index.ts:7]: Actually @viceice I disagree because you can (and in the real world projects **you will**) have the possibility to [include other hcl](https://terragrunt.gruntwork.io/docs/getting-started/configuration/#configuration-parsing-order) files (which don't need to be called terragrunt.hcl) but those files still need to comply with the Terragrunt structure.

So you can have something like this:

- main.hcl

```hcl
terraform {
  source = "github.com/bgdanix/tfrepo//terraform/modules/moduleone?ref=0.0.9"
}

inputs = {
  name           = "mysql_${local.env}"
  instance_class = "db.t2.micro"

  allocated_storage = 20
  storage_type      = "standard"

  master_username = "admin"
  # TODO: To avoid storing your DB password in the code, set it as the environment variable TF_VAR_master_password
}
```

- terragrunt.hcl

```hcl
include {
  path = find_in_parent_folders("main.hcl")
}

inputs = {
  instance_class = "db.t3.2xlarge"
}
```

So that your main file has the "default" values and when you actually go and run terragrunt ti will indeed only look at the `terragrunt.hcl` file but it will include whatever is in the `include` part and just overwrite what has changed, which in this case would be just changing the DB instance size.

Comment (rarkins) [reply lib/manager/terragrunt/index.ts:7]: What percentage of the world's HCL files would be terragrunt vs non-terragrunt?

Comment (rarkins) [reply lib/manager/terragrunt/index.ts:7]: > Actually @viceice I disagree because you can (and in the real world projects **you will**) have the possibility

What are you disagreeing with? @viceice wrote "A user can always override this if required" which is true.

Comment (bgdanix) [reply lib/manager/terragrunt/index.ts:7]: What I'm disagreeing with is we should not tie it to terragrunt.hcl files only, but on .hcl files, based on the example I've just given. And also in my case the users can override it to include other extensions, but at least catch any .hcl files

Comment (bgdanix) [reply lib/manager/terragrunt/index.ts:7]: > 
> What percentage of the world's HCL files would be terragrunt vs non-terragrunt?

Hmm, indeed, this would make more sense now since HCL is developed by [HashiCorp](https://www.hashicorp.com) and it can be used in any of its products, not only Terragrunt. Ok, I'm fine with the proposed change by @viceice

Comment (bgdanix) [reply lib/manager/terragrunt/index.ts:7]: I'll tag you once I've pushed the changed as I'd like to also indicate this limitation in the documentation

Comment (rarkins) [reply lib/manager/terragrunt/index.ts:7]: The question is not whether people do put terragunt configuration in alternate files, but about how many false hits we'll get if we default to `.hcl`. If terragrunt files represent only a small proportion of the world's `.hcl` files then Renovate users will end up doing a lot of wasted parsing. We could for example default the pattern `terragrunt.*\\.hcl$` which would mean any `.hcl` file which either has terragrunt in its filename or exists within any subfolder that includes `terragrunt` would be found

Comment (bgdanix) [reply lib/manager/terragrunt/index.ts]: Yes, you are right @rarkins , users will end up doing wasted parsing. Pushed the change now and also did two real tests by running the tool

**Comment by bgdanix** [comment]
@rarkins  I guess it can be merged now

**Comment by rarkins** [comment]
Thanks @bgdanix

**Comment by bgdanix** [comment]
Just for my info, do you guys now what is the background process for the documentation to be also updated here https://docs.renovatebot.com/modules/manager/ ? I'm asking as `terragrunt` is not yet showing up in there

**Comment by HonkingGoose** [comment]
@bgdanix I've made issue #7713 to track your question.
